### PR TITLE
Applied Refactoring 

### PR DESCRIPTION
**Pull request by BadhriNadh**
# Description

1. Variable 'v' does not have any meaning. So, renamed as 'includeReadNotifications',  'restrictedToParticipatingNotifications' and  ‘pollIntervalHeaderValue’.
2. Extracted a block to new method named 'getLatestUnReturnedThread',
3. Extracted a block to new method named 'waitUntilNextPollingTime'
4. Extracted a block to new method named 'updateThreadsAndMetadataFromGitHub'
5. Moved condition to a function that returns a boolean.
6. The ‘hashcode’ method in this class hierarchy [GHUser -> GHRepository] is overridden in a subclass and is only utilized in that subclass without being used in any other classes in the hierarchy. Therefore, the implementation of the hashcode method has been moved down to the subclass where it is actually used to simplify the parent class and make the code more modular.
7. The ‘getApiRoute’ method was present in multiple subclasses, but with slight implementation differences. To address this, I used the Template Method pattern and pulled-up the method to the parent class. This eliminated code duplication, simplified the codebase, and made it easier to maintain and update the common code. <--This is fixing a potential issue-->
8. Method ‘getApiTail’ has almost the same implementation in four classes: GHBlobBuilder, GHCommitBuilder, GHCommitComment, and GHTreeBuilder. It's also causing feature envy smell with GHRepository so moved these methods to GHRepository into one method. <--This is fixing a potential issue-->

# Before submitting a PR:

- [ ] Changes must not break binary backwards compatibility. If you are unclear on how to make the change you think is needed while maintaining backward compatibility, [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [ ] Add JavaDocs and other comments as appropriate. Consider including links in comments to relevant documentation on https://docs.github.com/en/rest .  
- [ ] Add tests that cover any added or changed code. This generally requires capturing snapshot test data. See [CONTRIBUTING.md](CONTRIBUTING.md) for details.
- [ ] Run `mvn -D enable-ci clean install site` locally. If this command doesn't succeed, your change will not pass CI.
- [ ] Push your changes to a branch other than `main`. You will create your PR from that branch.

# When creating a PR: 

- [ ] Fill in the "Description" above with clear summary of the changes. This includes:
  - [ ] If this PR fixes one or more issues, include "Fixes #<issue number>" lines for each issue. 
  - [ ] Provide links to relevant documentation on https://docs.github.com/en/rest where possible.
- [ ] All lines of new code should be covered by tests as reported by code coverage. Any lines that are not covered must have PR comments explaining why they cannot be covered. For example, "Reaching this particular exception is hard and is not a particular common scenario."
- [ ] Enable "Allow edits from maintainers".

### CODE REVIEW

File: src/main/java/org/kohsuke/github/GHObject.java [hunk 1]
```diff
@@ -190,4 +192,16 @@ public void append(StringBuffer buffer, String fieldName, Object value, Boolean
             super.append(buffer, fieldName, value, fullDetail);
         }
     };
+
+    public String getApiRoute(GHRepository owner) {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHObject.java:196]:
```suggestion
    protected String getApiRoute(GHRepository owner) {
```

File: src/main/java/org/kohsuke/github/GHArtifact.java [hunk 2]
```diff
@@ -122,15 +122,11 @@ public void delete() throws IOException {
...
     }
 
-    private String getApiRoute() {
-        if (owner == null) {
-            // Workflow runs returned from search to do not have an owner. Attempt to use url.
-            final URL url = Objects.requireNonNull(getUrl(), "Missing instance URL!");
-            return StringUtils.prependIfMissing(url.toString().replace(root().getApiUrl(), ""), "/");
-        }
+    @Override
+    protected String getApiRouteImpl() {
```
Comment (bitwiseman) [inline src/main/java/org/kohsuke/github/GHArtifact.java:129]: I don't love this `*Impl` pattern, but it makes sense here.

**Review by bitwiseman — CHANGES_REQUESTED** [review]
Generally approve of these refactors.  Some questions/requests.

**Comment by BadhriNadh** [comment]
Need a different approach for this api

**Pull request closed without merging**